#159987732 Users can get their reading stats

emmaadesile opened the pull request:
### What does this PR do?
Users should be able to get their reading stats.

### Description of tasks to be completed
- User's userId is logged to the `ArticleView` model when they view an article
- Users can get their reading stats.
- Reading stats include the following
  - The articles a user has read
  - The articles that a user has liked/disliked
  - The the most read category

### How can this be manually tested
- After cloning the app
- run `npm install` to install all the dependencies
- install and setup Postgresql
- create a `.env` file using the `.envsample` file
- on Postman, log in as a user and view an article at `/api/articles/:article-slug`
- Goto `/api/user-reading-stats` to view the reading stats of the user

### Screenshots
<img width="1105" alt="reading-stats" src="https://user-images.githubusercontent.com/26192691/45584621-da4c3000-b8ce-11e8-8094-62ee9b956ee7.png">


### Relevant Pivotal Tracker stories
[159987732](https://www.pivotaltracker.com/story/show/159987732)

emmaadesile committed:
fix uncaught error in category controller

emmaadesile committed:
ft(user-reading-stats): users can get their reading stats

- user's view should be logged everytime they get an article
- users should be able to see the number of articles read
- users should be able to get the articles they have read

[Delivers 159987732]

emmaadesile committed:
ft(users'-reading-stats): users can get their reading stats

- log user's view to the ArticleView model when they view an article
- users can get all the articles they have viewed
- users can get articles that users have liked
- users can get their most read category
- add more tests to Article model
- add tests to the ArticleView model
- sanitize category name before processing the request

[Delivers 159987732]

CEOehis commented on server/controllers/article.js:
```diff
@@ -211,10 +210,32 @@ class ArticleController {
+            if (!userView) {
+              ArticleView.create({ articleId, userId })
+                .then(() => res.status(201).json())
+                .catch(next);
```
The name of this method is `getSpecific`, I think it is not really descriptive. The method name should describe what it is getting. Also in lines`228-230` a resource is created which doesn't seem to tally with what I think the method should be doing.

I think you should have a separate method that creates an articleView.

emmaadesile replied:
The name `getSpecific` was the work of @ozezzy, but I'll fix it now.

emmaadesile replied:
Will work on it now

CEOehis commented on server/controllers/readingStatsController.js:
```diff
@@ -0,0 +1,80 @@
+              })
+              .catch(error => error);
+          })
+          .catch(error => error);
```
It is good that you have caught this error, but just returning the error without handling it is not as user experience friendly as handling it. You can instead, call next with the error and let the general error handler defined in the apps `index.js` handle it or handle the errors as you did in line 76.

This also applies to line 72.

emmaadesile replied:
Alright. I'll fix it now.

CEOehis commented on server/utils/modeValue.js:
```diff
@@ -0,0 +1,6 @@
+  - words.filter(v => v === b).length).pop()
+);
+
+export default mode;
```
Nice work on writing this utility method using functional Javascript style. I think you can make readers of your code understand what the utility does by using  JSDoc comments and also using more descriptive variable names for the variables in your array methods.

emmaadesile replied:
Good one. Duly noted. I'll make the changes now

emmaadesile committed:
ft(user-reading-stats):

- refactor mode function to use more descriptive variables
- add jsdoc comment to mode util function
- change getSpecific method name to getArticle in articles controller
- change getAll method name to getAllArticles in articles controller

[Delivers 159987732]

emmaadesile committed:
ft(user-reading-stats): fix failing travis build

[Delivers 159987732]

emmaadesile committed:
ft(user-reading-stats): fix failing travis build

[Delivers 159987732]

emekafredy commented on server/tests/controllers/category.test.js:
```diff
@@ -494,5 +472,22 @@ describe('Categorizes articles', () => {
+          expect(res).to.have.status(200);
+          expect(res.body.status).to.equal('success');
+          expect(res.body).to.have.property('category');
+          expect(res.body.category).to.have.property('createdAt');
```
It is always better to use object destructuring(lines 485 - 488). Also applies for lines 72 - 74 of your readingstats test file, lines 89 & 90.

emmaadesile replied:
I'll do that now.

emekafredy reviewed (changes requested):
Please, also update the swagger doc with the new endpoint(s) you defined.

emmaadesile committed:
ft(user-reading-stats): update swagger documentation

- update swagger documentation
- change the route for adding articles to a category

[Delivers 159987732]

aknwosu commented on server/controllers/categoryController.js:
```diff
@@ -38,7 +39,7 @@ class CategoryController {
     const name = req.body.name.trim();
-    const newCategory = `${name.substr(0, 1).toUpperCase()}${name.slice(1).toLowerCase()}`;
+    const newCategory = sanitizeString(name);
     if (newCategory === '') {
```
if (!newCategory) {

aknwosu commented on server/controllers/readingStatsController.js:
```diff
@@ -0,0 +1,109 @@
+                    mostReadCategory = mode(names);
+                    return res.status(200).json({
+                      status: 'success',
+                      articlesRead: articles.length === 0 ? 'You have not read any article' : articles,
```
No articles found

aknwosu commented on server/controllers/readingStatsController.js:
```diff
@@ -0,0 +1,109 @@
+                      articlesRead: articles.length === 0 ? 'You have not read any article' : articles,
+                      numberOfArticlesRead: articleCount.length,
+                      articleReactions: reaction.length === 0 ? 'You have not liked/disliked any article' : reaction,
+                      mostReadCategory
```
No reactions found

aknwosu commented on server/controllers/readingStatsController.js:
```diff
@@ -0,0 +1,109 @@
+   * @param  {string} articleId body of the user's request
+   * @memberof ReadingStatsController
+   */
+  static saveArticleView(req, res, next) {
```
postArticleViewHistory

aknwosu commented on server/models/articleview.js:
```diff
@@ -0,0 +1,13 @@
+module.exports = (sequelize, DataTypes) => {
+  const ArticleView = sequelize.define('ArticleView', {
```
articleViewHistory

emmaadesile committed:
ft(user-reading-stats): users should be able to get their reading stats

- change the name of saveArticleView method to postArticleHistory
- change the name of ArticleView model to ArticleViewHistory

[Delivers 159987732]

emmaadesile committed:
Fix merge conflicts in article.js file

emmaadesile committed:
ft(swagger-documentation): fix swagger docummentation error

[Fixes 159987732]

emmaadesile committed:
fix hound errors
